skaha/src/main/java/org/opencadc/skaha/utils/PosixHelper.java

```diff
@@ -17,15 +45,94 @@ public static String uidMapping(final PosixPrincipal posixPrincipal) {
+    static boolean posixMappingSecretRequiresUpdate() throws Exception {
+        final JSONObject jsonObject = CommandExecutioner.getSecretData(PosixHelper.POSIX_MAPPINGS_SECRET_NAME, K8SUtil.getWorkloadNamespace());
+        if (jsonObject.has(PosixHelper.POSIX_MAPPING_SECRET_LAST_MOD_KEY)) {
+            final String lastModDateString = CommonUtils.decodeBase64(jsonObject.getString(PosixHelper.POSIX_MAPPING_SECRET_LAST_MOD_KEY));
+            final Date lastModDate = DateUtil.getDateFormat(DateUtil.ISO8601_DATE_FORMAT_LOCAL, DateUtil.UTC).parse(lastModDateString);
+            final long currentAgeInMilliseconds = PosixHelper.currentDateUTC().getTime() - lastModDate.getTime();
+            LOGGER.debug("current age of " + PosixHelper.POSIX_MAPPINGS_SECRET_NAME + " is " + currentAgeInMilliseconds + " milliseconds");
+            return currentAgeInMilliseconds >= PosixHelper.ONE_MINUTE_MS;
+        } else {
+            return true;
+        }
+    }
+
+    public static void ensurePosixMappingSecret(final PosixMapperClient posixMapperClient) throws Exception {
+        final String posixMappingSecretName = PosixHelper.POSIX_MAPPINGS_SECRET_NAME;
+
+        if (PosixHelper.posixMappingSecretRequiresUpdate()) {
+            final Path holdingDir = Files.createTempDirectory(Path.of(System.getProperty("java.io.tmpdir")), posixMappingSecretName);
+            final File uidMappingFile = new File(holdingDir.toString(), "uidmap.txt");
+            final File gidMappingFile = new File(holdingDir.toString(), "gidmap.txt");
+            final String lastModDateString = DateUtil.getDateFormat(DateUtil.ISO8601_DATE_FORMAT_LOCAL, DateUtil.UTC).format(PosixHelper.currentDateUTC());
+
+            try (final FileWriter uidMappingWriter = new FileWriter(uidMappingFile)) {
+                uidMappingWriter.write(getUserEntries(posixMapperClient));
+                uidMappingWriter.flush();
+            }
+
+            try (final FileWriter gidMappingWriter = new FileWriter(gidMappingFile)) {
+                gidMappingWriter.write(getGroupEntries(posixMapperClient));
+                gidMappingWriter.flush();
+            }
+
+            // create new secret, or update the existing one.
+            final String[] createSecretCommand = new String[] {
+                "kubectl", "--namespace", K8SUtil.getWorkloadNamespace(), "create", "secret", "generic",
+                posixMappingSecretName,
+                "--from-file=" + uidMappingFile.getAbsolutePath(),
+                "--from-file=" + gidMappingFile.getAbsolutePath(),
+                "--from-literal=" + PosixHelper.POSIX_MAPPING_SECRET_LAST_MOD_KEY + "=" + lastModDateString,
+                "-o",
+                "yaml",
+                "--dry-run=client",
+                "|",
+                "kubectl",
+                "apply",
+                "-f",
+                "-"
+            };
+
+            final String createResult = CommandExecutioner.executeInShell(createSecretCommand, false);
```

Comment:
Curious as to why we need this to be wrapped in sh?

Reply:
Because of the pipe (`|`) in the command.